File: src/main/java/com/company/apimgmt/service/RelationshipService.java
package com.company.apimgmt.service;

import com.company.apimgmt.dto.CreateRelationshipRequest;
import com.company.apimgmt.dto.RelationshipDTO;
import com.company.apimgmt.dto.UpdateRelationshipRequest;
import com.company.apimgmt.entity.RelationshipEntity;
import com.company.apimgmt.enums.EntityType;
import com.company.apimgmt.exception.ResourceNotFoundException;
import com.company.apimgmt.exception.ValidationException;
import com.company.apimgmt.repository.ApiRepository;
import com.company.apimgmt.repository.EndpointRepository;
import com.company.apimgmt.repository.RelationshipRepository;
import com.company.apimgmt.repository.SystemRepository;
import io.smallrye.mutiny.Uni;
import io.vertx.core.json.JsonObject;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@ApplicationScoped
public class RelationshipService {
    
    @Inject
    RelationshipRepository relationshipRepository;
    
    @Inject
    SystemRepository systemRepository;
    
    @Inject
    ApiRepository apiRepository;
    
    @Inject
    EndpointRepository endpointRepository;
    
    public Uni<RelationshipDTO> createRelationship(CreateRelationshipRequest request) {
        // Validate caller exists
        return validateEntity(request.callerType, request.callerId, "Caller")
            .onItem().transformToUni(callerName -> 
                // Validate callee exists
                validateEntity(request.calleeType, request.calleeId, "Callee")
                    .onItem().transformToUni(calleeName -> 
                        // Validate endpoint exists and belongs to callee
                        validateEndpoint(request.endpointId, request.calleeType, request.calleeId)
                            .onItem().transformToUni(endpoint -> {
                                // Create relationship entity
                                RelationshipEntity entity = new RelationshipEntity();
                                entity.id = UUID.randomUUID();
                                entity.callerType = request.callerType;
                                entity.callerId = request.callerId;
                                entity.calleeType = request.calleeType;
                                entity.calleeId = request.calleeId;
                                entity.endpointId = request.endpointId;
                                entity.authType = request.authType;
                                entity.authConfig = request.authConfig != null ? 
                                    new JsonObject(request.authConfig) : null;
                                entity.description = request.description;
                                entity.createdAt = LocalDateTime.now();
                                entity.updatedAt = LocalDateTime.now();
                                
                                return relationshipRepository.create(entity)
                                    .onItem().transform(created -> 
                                        toDTO(created, callerName, calleeName, 
                                            endpoint.path, endpoint.httpMethod)
                                    );
                            })
                    )
            );
    }
    
    public Uni<RelationshipDTO> getRelationshipById(UUID id) {
        return relationshipRepository.findById(id)
            .onItem().transformToUni(entity -> {
                if (entity == null) {
                    return Uni.createFrom().failure(
                        new ResourceNotFoundException("Relationship", id)
                    );
                }
                return enrichRelationshipDTO(entity);
            });
    }
    
    public Uni<List<RelationshipDTO>> getAllRelationships() {
        return relationshipRepository.findAll()
            .onItem().transformToUni(entities -> {
                List<Uni<RelationshipDTO>> dtoUnis = entities.stream()
                    .map(this::enrichRelationshipDTO)
                    .collect(Collectors.toList());
                
                return Uni.combine().all().unis(dtoUnis)
                    .combinedWith(dtos -> 
                        dtos.stream()
                            .map(dto -> (RelationshipDTO) dto)
                            .collect(Collectors.toList())
                    );
            });
    }
    
    public Uni<List<RelationshipDTO>> getRelationshipsByCaller(EntityType type, UUID id) {
        return relationshipRepository.findByCaller(type, id)
            .onItem().transformToUni(entities -> {
                List<Uni<RelationshipDTO>> dtoUnis = entities.stream()
                    .map(this::enrichRelationshipDTO)
                    .collect(Collectors.toList());
                
                return Uni.combine().all().unis(dtoUnis)
                    .combinedWith(dtos -> 
                        dtos.stream()
                            .map(dto -> (RelationshipDTO) dto)
                            .collect(Collectors.toList())
                    );
            });
    }
    
    public Uni<List<RelationshipDTO>> getRelationshipsByCallee(EntityType type, UUID id) {
        return relationshipRepository.findByCallee(type, id)
            .onItem().transformToUni(entities -> {
                List<Uni<RelationshipDTO>> dtoUnis = entities.stream()
                    .map(this::enrichRelationshipDTO)
                    .collect(Collectors.toList());
                
                return Uni.combine().all().unis(dtoUnis)
                    .combinedWith(dtos -> 
                        dtos.stream()
                            .map(dto -> (RelationshipDTO) dto)
                            .collect(Collectors.toList())
                    );
            });
    }
    
    public Uni<RelationshipDTO> updateRelationship(UUID id, UpdateRelationshipRequest request) {
        return relationshipRepository.findById(id)
            .onItem().transformToUni(entity -> {
                if (entity == null) {
                    return Uni.createFrom().failure(
                        new ResourceNotFoundException("Relationship", id)
                    );
                }
                
                // Determine which fields to validate
                EntityType newCallerType = request.callerType != null ? request.callerType : entity.callerType;
                UUID newCallerId = request.callerId != null ? request.callerId : entity.callerId;
                EntityType newCalleeType = request.calleeType != null ? request.calleeType : entity.calleeType;
                UUID newCalleeId = request.calleeId != null ? request.calleeId : entity.calleeId;
                UUID newEndpointId = request.endpointId != null ? request.endpointId : entity.endpointId;
                
                // Validate caller if changed
                Uni<String> callerValidation = 
                    (request.callerType != null || request.callerId != null) ?
                    validateEntity(newCallerType, newCallerId, "Caller") :
                    getEntityName(entity.callerType, entity.callerId);
                
                return callerValidation.onItem().transformToUni(callerName -> {
                    // Validate callee if changed
                    Uni<String> calleeValidation = 
                        (request.calleeType != null || request.calleeId != null) ?
                        validateEntity(newCalleeType, newCalleeId, "Callee") :
                        getEntityName(entity.calleeType, entity.calleeId);
                    
                    return calleeValidation.onItem().transformToUni(calleeName -> {
                        // Validate endpoint if changed
                        Uni<Void> endpointValidation = 
                            request.endpointId != null ?
                            validateEndpoint(newEndpointId, newCalleeType, newCalleeId)
                                .onItem().transform(e -> null) :
                            Uni.createFrom().voidItem();
                        
                        return endpointValidation.onItem().transformToUni(v -> {
                            // Update entity fields
                            if (request.callerType != null) {
                                entity.callerType = request.callerType;
                            }
                            if (request.callerId != null) {
                                entity.callerId = request.callerId;
                            }
                            if (request.calleeType != null) {
                                entity.calleeType = request.calleeType;
                            }
                            if (request.calleeId != null) {
                                entity.calleeId = request.calleeId;
                            }
                            if (request.endpointId != null) {
                                entity.endpointId = request.endpointId;
                            }
                            if (request.authType != null) {
                                entity.authType = request.authType;
                            }
                            if (request.authConfig != null) {
                                entity.authConfig = new JsonObject(request.authConfig);
                            }
                            if (request.description != null) {
                                entity.description = request.description;
                            }
                            entity.updatedAt = LocalDateTime.now();
                            
                            return relationshipRepository.update(entity)
                                .onItem().transformToUni(this::enrichRelationshipDTO);
                        });
                    });
                });
            });
    }
    
    public Uni<Void> deleteRelationship(UUID id) {
        return relationshipRepository.findById(id)
            .onItem().transformToUni(entity -> {
                if (entity == null) {
                    return Uni.createFrom().failure(
                        new ResourceNotFoundException("Relationship", id)
                    );
                }
                
                return relationshipRepository.delete(id)
                    .onItem().transform(deleted -> null);
            });
    }
    
    private Uni<String> validateEntity(EntityType type, UUID id, String role) {
        if (type == EntityType.SYSTEM) {
            return systemRepository.findById(id)
                .onItem().transform(entity -> {
                    if (entity == null) {
                        throw new ResourceNotFoundException(role + " System", id);
                    }
                    return entity.name;
                });
        } else {
            return apiRepository.findById(id)
                .onItem().transform(entity -> {
                    if (entity == null) {
                        throw new ResourceNotFoundException(role + " API", id);
                    }
                    return entity.name;
                });
        }
    }
    
    private Uni<String> getEntityName(EntityType type, UUID id) {
        if (type == EntityType.SYSTEM) {
            return systemRepository.findById(id)
                .onItem().transform(entity -> entity != null ? entity.name : null);
        } else {
            return apiRepository.findById(id)
                .onItem().transform(entity -> entity != null ? entity.name : null);
        }
    }
    
    private Uni<com.company.apimgmt.entity.EndpointEntity> validateEndpoint(
            UUID endpointId, EntityType calleeType, UUID calleeId) {
        return endpointRepository.findById(endpointId)
            .onItem().transform(endpoint -> {
                if (endpoint == null) {
                    throw new ResourceNotFoundException("Endpoint", endpointId);
                }
                
                // Verify endpoint belongs to the callee
                if (calleeType == EntityType.API) {
                    if (!endpoint.apiId.equals(calleeId)) {
                        throw new ValidationException(
                            "Endpoint does not belong to the specified API"
                        );
                    }
                } else {
                    // For SYSTEM callee, we need to verify the endpoint's API belongs to the system
                    // This will be checked via API lookup
                    return endpoint;
                }
                
                return endpoint;
            })
            .onItem().transformToUni(endpoint -> {
                if (calleeType == EntityType.SYSTEM) {
                    return apiRepository.findById(endpoint.apiId)
                        .onItem().transform(api -> {
                            if (api == null || !api.systemId.equals(calleeId)) {
                                throw new ValidationException(
                                    "Endpoint does not belong to an API in the specified System"
                                );
                            }
                            return endpoint;
                        });
                }
                return Uni.createFrom().item(endpoint);
            });
    }
    
    private Uni<RelationshipDTO> enrichRelationshipDTO(RelationshipEntity entity) {
        return getEntityName(entity.callerType, entity.callerId)
            .onItem().transformToUni(callerName -> 
                getEntityName(entity.calleeType, entity.calleeId)
                    .onItem().transformToUni(calleeName -> 
                        endpointRepository.findById(entity.endpointId)
                            .onItem().transform(endpoint -> 
                                toDTO(entity, callerName, calleeName,
                                    endpoint != null ? endpoint.path : null,
                                    endpoint != null ? endpoint.httpMethod : null)
                            )
                    )
            );
    }
    
    private RelationshipDTO toDTO(RelationshipEntity entity, String callerName, String calleeName,
                                  String endpointPath, com.company.apimgmt.enums.HttpMethod endpointMethod) {
        Map<String, Object> authConfigMap = null;
        if (entity.authConfig != null) {
            authConfigMap = new HashMap<>(entity.authConfig.getMap());
        }
        
        return new RelationshipDTO(
            entity.id,
            entity.callerType,
            entity.callerId,
            callerName,
            entity.calleeType,
            entity.calleeId,
            calleeName,
            entity.endpointId,
            endpointPath,
            endpointMethod,
            entity.authType,
            authConfigMap,
            entity.description,
            entity.createdAt,
            entity.updatedAt
        );
    }
}

File: apimgmt/services/relationship_service.py
from typing import List, Optional
import uuid
from datetime import datetime

from apimgmt.repositories.relationship_repository import RelationshipRepository
from apimgmt.repositories.api_repository import ApiRepository
from apimgmt.repositories.endpoint_repository import EndpointRepository
from apimgmt.services.audit_service import AuditLogService
from apimgmt.models.relationship import Relationship
from apimgmt.schemas.relationship import CreateRelationshipRequest, UpdateRelationshipRequest, RelationshipDTO
from apimgmt.exceptions import ResourceNotFoundException
from apimgmt.utils.audit_decorator import audit_log
from apimgmt.models.audit import OperationType, ResourceType


class RelationshipService:
    def __init__(self, relationship_repository: RelationshipRepository, api_repository: ApiRepository, endpoint_repository: EndpointRepository, audit_service: Optional[AuditLogService] = None):
        self.relationship_repository = relationship_repository
        self.api_repository = api_repository
        self.endpoint_repository = endpoint_repository
        self.audit_service = audit_service
    
    @audit_log(
        operation_type=OperationType.CREATE,
        resource_type=ResourceType.RELATIONSHIP
    )
    def create_relationship(self, request: CreateRelationshipRequest) -> RelationshipDTO:
        """创建调用关系"""
        # 验证端点是否存在
        if request.endpoint_id:
            endpoint = self.endpoint_repository.find_by_id(request.endpoint_id)
            if not endpoint:
                raise ResourceNotFoundException("Endpoint", request.endpoint_id)
        
        # 创建调用关系
        relationship = Relationship(
            id=str(uuid.uuid4()),
            caller_type=request.caller_type,
            caller_id=request.caller_id,
            callee_type=request.callee_type,
            callee_id=request.callee_id,
            endpoint_id=request.endpoint_id,
            auth_type=request.auth_type,
            auth_config=request.auth_config,
            description=request.description,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        created_relationship = self.relationship_repository.create(relationship)
        
        return RelationshipDTO.model_validate(created_relationship)
    
    def get_relationship_by_id(self, relationship_id: uuid.UUID) -> RelationshipDTO:
        """根据ID获取调用关系"""
        relationship = self.relationship_repository.find_by_id(relationship_id)
        if not relationship:
            raise ResourceNotFoundException("Relationship", relationship_id)
        return RelationshipDTO.model_validate(relationship)
    
    def get_all_relationships(self) -> List[RelationshipDTO]:
        """获取所有调用关系"""
        relationships = self.relationship_repository.find_all()
        return [RelationshipDTO.model_validate(relationship) for relationship in relationships]
    
    def get_relationships_by_caller(self, caller_type: str, caller_id: uuid.UUID) -> List[RelationshipDTO]:
        """根据调用方获取调用关系"""
        relationships = self.relationship_repository.find_by_caller(caller_type, caller_id)
        return [RelationshipDTO.model_validate(relationship) for relationship in relationships]
    
    def get_relationships_by_callee(self, callee_type: str, callee_id: uuid.UUID) -> List[RelationshipDTO]:
        """根据被调用方获取调用关系"""
        relationships = self.relationship_repository.find_by_callee(callee_type, callee_id)
        return [RelationshipDTO.model_validate(relationship) for relationship in relationships]
    
    @audit_log(
        operation_type=OperationType.UPDATE,
        resource_type=ResourceType.RELATIONSHIP,
        resource_id_param="relationship_id"
    )
    def update_relationship(self, relationship_id: uuid.UUID, request: UpdateRelationshipRequest) -> RelationshipDTO:
        """更新调用关系"""
        # 查找调用关系
        relationship = self.relationship_repository.find_by_id(relationship_id)
        if not relationship:
            raise ResourceNotFoundException("Relationship", relationship_id)
        
        # 更新字段
        if request.caller_type:
            relationship.caller_type = request.caller_type
        if request.caller_id:
            relationship.caller_id = request.caller_id
        if request.callee_type:
            relationship.callee_type = request.callee_type
        if request.callee_id:
            relationship.callee_id = request.callee_id
        if request.endpoint_id:
            # 验证端点是否存在
            endpoint = self.endpoint_repository.find_by_id(request.endpoint_id)
            if not endpoint:
                raise ResourceNotFoundException("Endpoint", request.endpoint_id)
            relationship.endpoint_id = request.endpoint_id
        if request.auth_type is not None:
            relationship.auth_type = request.auth_type
        if request.auth_config is not None:
            relationship.auth_config = request.auth_config
        if request.description is not None:
            relationship.description = request.description
        relationship.updated_at = datetime.utcnow()
        
        updated_relationship = self.relationship_repository.update(relationship)
        
        return RelationshipDTO.model_validate(updated_relationship)
    
    @audit_log(
        operation_type=OperationType.DELETE,
        resource_type=ResourceType.RELATIONSHIP,
        resource_id_param="relationship_id"
    )
    def delete_relationship(self, relationship_id: uuid.UUID) -> None:
        """删除调用关系"""
        # 验证调用关系是否存在
        relationship = self.relationship_repository.find_by_id(relationship_id)
        if not relationship:
            raise ResourceNotFoundException("Relationship", relationship_id)
        
        # 删除调用关系
        self.relationship_repository.delete(relationship_id)
